### PR TITLE
Add a flow for returning data early from Mutation#authorized?

### PR DESCRIPTION
**Pull request by rmosolgo**
As pointed out by @cainlevy in #1734 , the new mutation auth doesn't support errors-as-data very well, but it definitely _should_, because errors-as-data is nice. 

How to do it? I can think of a few implementations: 

- Somehow use the return value of `.authorized?` (this is how before-save callbacks used to work, if they returned `false`, then save was prevented. I think this was changed to [`throw(:abort)`](https://blog.bigbinary.com/2016/02/13/rails-5-does-not-halt-callback-chain-when-false-is-returned.html)
- Use `throw :early_return, response`. This has the advantage of being plain Ruby, but it's not a very familiar feature. Also, it halts program flow, which is convenient, since later code will be irrelevant. 
- Use a normal method, as shown here. It's abstract, which is nice, but doesn't halt flow. This is like Rails controller `render` methods.
- Use a method that calls `throw` under the hood (like sinatra's `halt`, I think). It's nice that it halts flow, but maybe too magical? 

Any thoughts? 

TODO: 

- [ ] before_prepare should get the same treatment, be renamed to a `?` method.

### CODE REVIEW

**Comment by cainlevy** [comment]
Thoughts on creating an official GraphQL::Result class? I imagine it's no small undertaking and would trickle out to many bits of the API, but it seems like the closest way to represent errors as data.

One option might be to always expect a Result back from `.authorized?`. The developer would need to indicate when it's a "continue" Result and when it's a "halt" result. Honestly this just feels like a Promise.

Another option would be to continue for any truthy value that is _not_ a `GraphQL::Result` object. In that case, `.authorized?` would need to return `GraphQL::Result.new(data)` when it wants to halt resolution with error data.

**Comment by rmosolgo** [comment]
>  any truthy value that is _not_

Oh, that's a nice direction, something similar would be: 

- Continue on `true`
- Halt on:
  - `false`, with no data at all 
  - a non-`true` value, using that value as the response 

In that way, returning a hash would Just Work :tm:

**Comment by cainlevy** [comment]
I'm inclined to caution in Ruby when there's a difference between `true` and truthy (or between `false` and falsey). I've had co-workers who thought it was an anti-pattern to use `!!` for casting a result to a strict boolean.

**Comment by rmosolgo** [comment]
Oh yeah, I saw a rubocop against it once! 🤷‍♂️

**Comment by rmosolgo** [comment]
I updated this to use multiple return for errors-as-data, eg 

```ruby
return false, { errors: [...] }
```

I also updated the guides to give a sense of it. How does that look to you?

File: lib/graphql/schema/resolver.rb [hunk 1]
```diff
@@ -210,6 +222,23 @@ def load_application_object_failed(err)
...
+      # Call this during loading or authorization to
+      # skip {#resolve} and return data instead of top-level errors.
+      #
+      # It doesn't halt the flow of the code so be sure to return
+      # after calling it, or combine it with `return return_with(...)`.
+      #
+      #
+      # @param response [Object] The value to respond with
+      # @return [void]
+      def return_with(response)
```
Comment (cainlevy) [inline lib/graphql/schema/resolver.rb]: was this a previous idea? if it's still on the table, i think `resolve_with` might fit the GraphQL nomenclature nicely.

Comment (rmosolgo) [reply lib/graphql/schema/resolver.rb]: oops, oh yeah, it can be removed now since `return` is used instead

**Pull request merged**